### PR TITLE
fix(persist): incorrect deserialize return type

### LINKED ISSUE #465: persist middleware deserialize typings incorrect.
I am trying to use a custom `deserialize` function in order to avoid crashes/mismatches between runtime and typescript types. However, the typings of `persist` seem to be incorrect, as TypeScript complains that the setter function attached to the state is not returned from the `deserialize` function:

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAYygUwIY2XAZlCI4BEAXgK4DOMqAdgCYEDcAUKJLHAN5xjJRnAVwAvtlz5i5SrQD0IYDRoAbZAHdUKBs3DR4AKjioycAFJkIVEXkJYwAWhhkpC4ACMpJsxpbaO2BRGUANHByyFQwwDAAnkIWYtZ2Dk6uWCRUCOEeTF5segZwAKKxVrb2ji5S+REAFjyeWjn6hsAQRQTNCRqMCGYCACoQANahALKhJPkAHmDUNMg0AMqUGMMQswpwALzBEAB0UdwAFOyMcMFk-UNUAHIQGAAiyGRIwGAZVJPTtHMAXNs7zhAIEpqAETmcLiMxh8Zj8-gCgWgqKDBABKJhdHrwAAKPD4FDmADVUAoSJgtgAeXq9SLcODICYYWhNXbU7jk6iRAB8nIOFHQyBWa1+VJpyBRm05YOae1FRzBpwAbrjmlRftKqCQQM4eKDTqc+RhfgaBatkApdUI0YxGPtMBCqKMNdCvgslia1qyyX9PQB5LDksG2iBYOD2x3jKYw138wVmxic9G20ODSFOyMuxb8zbJy7h52zaPLU0KT1wABkHDBZGQMAAkucU9dbsgHk8oC83vnYQd+F2aL94cCqOKNpy4AqIHImPqa-Ww1D0wXfj2yH2B4ChyOxxOp4xBOjulQBLMsKgSAoYPO058C5mi0K4AAlNA0MwKSKUxt5xdzO-us2emOGxgscergo2Nz3I8zyvCqa5wDAUCkhapz8FeEY3rCiHIWC+7WoeAjkMg6F9n+2ZIGgGCfrmC6YYWyA8mC3C8PwMAHPKcC8jWW6caBYFwDsgknmeF4kT+9GxuaHEznWDaXJBLbQe2sFmPBK72gprYwZ24lbtJerVmxnBoRBzZacpOl0ZaKEGbOcmphhUbLiZNHXlGen8bZRnga5jkutZHGojZfFgVQqAgMgvwEDAX60VGf4EDZpxSFIcDKJgszVu2xLAEQmAwDU47EqSwTmKgcBkKgWCYKo0TfCi+mZTwwA5XlvxYH4yjsZ5pzuFQOzTLwyBJXq+Q7AgVSoMAVAAOoHDiLH4jQRIksgBxiXRf6SSiOyzN0swoiNpxjVgQI0N1PWnAcPFypdYFKixZi-AADEd-HGr8nCCbtyCnuel6xW5GZupJQhvacQX6WBIRhBEkRQ3ADWeYdgVgg1aJAA

I guess the correct solution would be to have a helper type that recursively removes function properties from the state type.

### PR DESCRIPTION
**Pull request by AnatoleLucet**
Closes #465 

@n1ru4l can you try this out with the codesandbox below to see if it fixes your issue? :pray: :slightly_smiling_face:

### CODE REVIEW

File: src/middleware.ts [hunk 1]
```diff
@@ -156,6 +156,18 @@ export const combine =
       )
     )
 
+type RemoveMethods<T> = Pick<
```
Comment (AnatoleLucet) [inline src/middleware.ts:159]: btw, it's always a joy to get headache while working with ts generics :upside_down_face:

**Review by dai-shi — COMMENTED** [review]
I wonder if this is a right solution. A custom deserialize may support deserializing methods somehow. We don't deep merge, so recursive RemoveMethods seems incorrect (thus type assertion).

Can we solve it with generics?
```ts
type StateStorage<PersistedState extends Partial<S> = S> = {
  ...
  deserialize?: (str: string) => StorageValue<PersistedState> | Promise<StorageValue<PersistedState>>
```

**Comment by AnatoleLucet** [comment]
Yeah, that's a better idea.
Though I'm not sure to understand why you want to add this generic & method on `StateStorage`. Did you mean `PersistOptions`?

**Comment by dai-shi** [comment]
Ah, that's my mistake. Yes, `PersistOptions`.

**Comment by AnatoleLucet** [comment]
Only thing with a shallow partial is, what about cases like #457? Maybe a deep partial could be ok in this case?

**Comment by dai-shi** [comment]
Sorry if I don't get it correctly, but does it only apply if a user pass custom `merge` option? If the merge function is fully typed, it should just work. Or is it about something other than `merge`?

**Comment by AnatoleLucet** [comment]
I doesn't have anything to do with `merge` :wink: 

Here's an example:
```ts
const unstorableMethod = () => {}

const useStore = create(
  persist(
    () => ({
      count: 0,
      unstorableMethod,
      object: { anotherCount: 0, anotherUnstorableMethod: unstorableMethod },
    }),
    {
      name: 'test-storage',
      deserialize: () => {
        return {
          state: {
            object: {
              anotherCount: 0, // error, `anotherUnstorableMethod` is not optional in this object
            }
          },
          version: 0,
        }
      },
    }
  )
)
```
image

**Comment by dai-shi** [comment]
To this to work with shallow, the state type must be
```ts
type TState = {
  count: number
  unstorableMethod: () => unknown
  object: {
    anotherCount: number
    anotherUnstorableMethod?: () => unknown
  }
}
```
So the type error without it is correct, I suppose.

BTW, `merge` option is only effective with "sync". I guess we would want to have the same behavior with "async".

**Comment by AnatoleLucet** [comment]
I pushed the changes

> BTW, merge option is only effective with "sync". I guess we would want to have the same behavior with "async".

Not sure what you mean. Do you have an example?

**Review by dai-shi — APPROVED** [review]
The change looks good.

**Comment by dai-shi** [comment]
> I pushed the changes
> 
> > BTW, merge option is only effective with "sync". I guess we would want to have the same behavior with "async".
> 
> Not sure what you mean. Do you have an example?

In your code example https://github.com/pmndrs/zustand/pull/468#issuecomment-869467669,
if we pass`merge=deepMerge`,
- with sync storage, anotherUnstorableMethod is preserved
- with async storage, anotherUnstorableMethod is not preserved (removed)

Is it correct?

**Comment by AnatoleLucet** [comment]
I don't really understand why it wouldn't, but I tried modifying a test, and it seems to work just like with a sync storage. What makes you think that `anotherUnstorableMethod` wouldn't be preserved?

<details><summary>The test</summary>
<p>

```tsx
  it('gives the merged state to onRehydrateStorage', async () => {
    const onRehydrateStorageSpy = jest.fn()

    const storage = {
      getItem: async () =>
        JSON.stringify({
          state: { count: 1 },
          version: 0,
        }),
      setItem: () => {},
    }

    const unstorableMethod = () => {}

    const isObject = (o: Object) => o?.toString() === '[object Object]'

    const deepMerge = <T extends Object, S extends Object>(
      target: T,
      source?: S
    ): T & S => {
      if (!source) return target as T & S

      const mergedObject: Partial<T & S> = {}
      const mergedEntries: [keyof T & S, (T & S)[keyof T & S]][] =
        Object.entries(target).concat(Object.entries(source)) as any

      mergedEntries.forEach(([key, value]) => {
        if (isObject(target[key]) && isObject(value)) {
          mergedObject[key] = deepMerge(target[key], value)
        } else {
          mergedObject[key] = value
        }
      })

      return mergedObject as T & S
    }

    const useStore = create(
      persist(
        () => ({
          count: 0,
          unstorableMethod,
          object: {
            anotherCount: 0,
            anotherUnstorableMethod: unstorableMethod,
          },
        }),
        {
          name: 'test-storage',
          getStorage: () => storage,
          onRehydrateStorage: () => onRehydrateStorageSpy,
          deserialize: () => {
            return {
              state: {
                count: 1,
                object: {
                  anotherCount: 0,
                },
              },
              version: 0,
            }
          },
          merge: (persistedState, currentState) =>
            deepMerge(currentState, persistedState),
        }
      )
    )

    function Counter() {
      const { count } = useStore()
      return <div>count: {count}</div>
    }

    const { findByText } = render(<Counter />)

    await findByText('count: 0')
    expect(onRehydrateStorageSpy).toBeCalledWith(
      {
        count: 1,
        unstorableMethod,
        object: { anotherCount: 0, anotherUnstorableMethod: unstorableMethod },
      },
      undefined
    )
  })
```

image

</p>
</details>

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/tree/test-persist-merge-nested
@AnatoleLucet A branch is pushed. Hope I don't mistake.

**Comment by AnatoleLucet** [comment]
Oh yeah, you're right. Nice catch! I forgot to make the storage async in the tests :sweat_smile: 
I'll have a look into it.

**Comment by dai-shi** [comment]
please merge master.

**Comment by AnatoleLucet** [comment]
done :slightly_smiling_face:

**Comment by dai-shi** [comment]
The change seems good (harmless), but I wonder if this alone solves @n1ru4l 's issue?

**Comment by AnatoleLucet** [comment]
I'm not familiar at all with fp-ts & io-ts, but it does seem to work by copy pasting @n1ru4l's code in this pr's csb
https://codesandbox.io/s/react-typescript-forked-ubpf8?file=/src/test.ts

**Comment by dai-shi** [comment]
That's great to know. Thanks for confirming!

**Comment by n1ru4l** [comment]
@AnatoleLucet awesome! Thank you for solving this so quick ❤️

**Pull request merged**